### ISSUE TITLE
The merged pull request on Feb 3 is breaking ~> slice_matrix

### ISSUE THREAD

**Issue by chaseleif** [blingenf/copydetect#36]
I could fix this issue by going back to the commit before the merge:
"only run CI on pull requests and master pushes" 62d4bfd

I have used copydetect multiple times before with no issues.

I had deleted the repo and installation files to save space, so when I re-cloned it I got the merged changes from Februrary.

I have tried varying the values in my config for "noise_threshold", "guarantee_threshold" and "display_threshold"

Each time I run it I get different indices in the KeyError but the same trace otherwise:
```
Traceback (most recent call last):
  File ".local/bin/copydetect", line 11, in <module>
    sys.exit(main())
  File ".local/lib/python3.6/site-packages/copydetect/__main__.py", line 117, in main
    detector.generate_html_report()
  File ".local/lib/python3.6/site-packages/copydetect/detector.py", line 637, in generate_html_report
    code_list = self.get_copied_code_list()
  File ".local/lib/python3.6/site-packages/copydetect/detector.py", line 598, in get_copied_code_list
    slices_test = self.slice_matrix[(y[idx], x[idx])][1]
KeyError: (9, 16)
```

In __detector.py__ (line numbers may be slightly varied from modification, some lines skipped).

There is a statement, `if (x[idx], y[idx]) in self.slice_matrix:`
Then there was just an else, assuming they would be in there reversed which was not happening.
I can get the script to complete and make my report.html by adding the else:
```
560     def get_copied_code_list(self):
594             if (x[idx], y[idx]) in self.slice_matrix:
595                 slices_test = self.slice_matrix[(x[idx], y[idx])][0]
596                 slices_ref = self.slice_matrix[(x[idx], y[idx])][1]
597             elif (y[idx], x[idx]) in self.slice_matrix:
598                 slices_test = self.slice_matrix[(y[idx], x[idx])][1]
599                 slices_ref = self.slice_matrix[(y[idx], x[idx])][0]
600             else:
601                 print(f'({x[idx]}, {y[idx]}) not in self.slice_matrix')
602                 slices_test = np.array([])
603                 slices_ref = np.array([])
```
By adding in the elif and the final else I can just get this warning:
```
.local/lib/python3.6/site-packages/copydetect/utils.py:167: UserWarning: empty slices array
  warnings.warn("empty slices array")
(11, 0) not in self.slice_matrix
```

I notice here that self.slice_matrix[] is only conditionally assigned:
```
485     def _comparison_loop(self):
515         for i, test_f in enumerate(tqdm(self.test_files,
517             for j, ref_f in enumerate(self.ref_files):
531                 else:
532                     overlap, (sim1, sim2), (slices1, slices2) = compare_files(
533                         self.file_data[test_f], self.file_data[ref_f]
534                     )
535                     comparisons[(test_f, ref_f)] = (i, j)
536                     if slices1.shape[0] != 0:
537                         self.slice_matrix[(i, j)] = [slices1, slices2]
```

I notice here that len(slices2[0]) is not checked, adding a check doesn't help but it seems odd:
```
100 def compare_files(file1_data, file2_data):
130     slices1 = get_copied_slices(idx1, file1_data.k)
131     slices2 = get_copied_slices(idx2, file2_data.k)
132     if len(slices1[0]) == 0:
133         return 0, (0,0), (np.array([]), np.array([]))
156     return token_overlap1, (similarity1,similarity2), (slices1,slices2)
```
`if len(slices1[0]) == 0 or len(slices2[0]) == 0:`

I saw in the commit history that changes happened to the slice_matrix, so reverted to the commit before the merge and now it is working again.

It wasn't very many lines that were changed, the problem should be in here: 16874d6

**Comment by chaseleif** [comment]
This can be fixed 1 of 2 ways.

First way to fix it, verify the indices exist:
```
560     def get_copied_code_list(self):
...
584         for idx in range(len(x)):
585             if ((x[idx], y[idx]) not in self.slice_matrix
586                     and (y[idx], x[idx]) not in self.slice_matrix):
587                 continue
```
Second way to fix it, don't conditionally add indices
original:
```
531                 else:
532                     overlap, (sim1, sim2), (slices1, slices2) = compare_files(
533                         self.file_data[test_f], self.file_data[ref_f]
534                     )
535                     comparisons[(test_f, ref_f)] = (i, j)
536                     if slices1.shape[0] != 0:
537                         self.slice_matrix[(i, j)] = [slices1, slices2]
```
new:
```
531                 else:
532                     overlap, (sim1, sim2), (slices1, slices2) = compare_files(
533                         self.file_data[test_f], self.file_data[ref_f]
534                     )
535                     comparisons[(test_f, ref_f)] = (i, j)
536                     self.slice_matrix[(i, j)] = [slices1, slices2]
```

**Comment by blingenf** [comment]
Thank you for reporting this and doing all the investigative work. I tinkered around a bit trying to replicate the issue and was able to cause a crash by providing both `test_dirs` and `ref_dirs`, with multiple items for one of the two. This might not be the exact cause of your issue, but the root cause is probably the same (the way the slice matrix is indexed appears to be incorrect). I'll look into this more and should hopefully have a fix for the root cause sometime soon.

**Comment by blingenf** [comment]
@chaseleif Would you mind checking if the new `fix/slice_mtx_indexing` branch fixes this issue?

The refactored slice matrix from February was written to avoid storing duplicate slice pairs, but makes the assumption that if a file is in both reference files and test files, it has the same index in both lists. However, there isn't any constraint actually forcing this to be the case. Because it's a dictionary instead of an actual matrix now, I just switched the keys from indexes to file names. That fixed the version of this issue I was able to replicate.

**Comment by chaseleif** [comment]
@blingenf Yes this is working for me now, thanks!

I have current sources each in their own subdirectory in a folder, like ./current/
I have previous sources the same way in ./old/
I have some provided example code loose in ./templates/
I have essentially:
```
test_directories = ['./current/' + name for name in listdir('current')]
reference_directories = "test_directories".copy()
reference_directories += ['./old/' + name for name in listdir('old')]
boilerplate_directories = ['./templates/']
extensions = ['c','cpp','h','hpp']
force_language = 'cpp'
```

**Comment by chaseleif** [comment]
I'm not sure if an OrderedDict may be worth it, or would work, but that could be a way to keep stable indices without storing filenames.

I'm not particularly worried about memory usage for my use-case, but if the dict were to grow large then not using filenames as keys could be good.

**Comment by chaseleif** [comment]
or mazbe something like
```
files = {}
i = 0
for name in testfiles:
  files[name] = i
  i += 1
for name in referencefiles:
  if name in files: continue
  files[name] = i
  i += 1
for name in boilerfiles:
  if name in files: continue
  files[name] = i
  i += 1
```
Then every file could have a unique index without storing duplicate names

**Comment by chaseleif** [comment]
You could store the ref and boiler startindex and drop the other name lists.
Just use 1 file "list", with like for i,name in enumerate(files) ...
Or maybe keeping it this way the OrderedDict could be useful...
```
files = OrderedDict()
for name in testfiles:
files[name] = None
# then when you assign it only use i
for i in range(refstart):
for i in range(refstart,boilerstart):
for i in range(boilerstart,len(files)):
```

**Comment by blingenf** [comment]
The fix is now on pypi as version 0.4.5. Feel free to reopen this issue if it comes up again with the latest version.

As far as memory goes -- the comparison matrix is typically sparse so I think the keys shouldn't be taking up too much memory relative to, e.g., the raw and filtered code files. The reason for the refactor that caused this issue was that the original code was storing N*M empty slice arrays regardless of how much similarity there was, which was a huge waste of memory. I do think storing a mapping of file name to index is a good idea though. That would also make it much easier to (for example) grab the similarity score between two specific files from the detector.
